pythonFiles/testing_tools/adapter/util.py

```diff
@@ -171,27 +168,79 @@ def fix_fileid(
 
 
 @contextlib.contextmanager
-def hide_stdio():
-    """Swallow stdout and stderr."""
-    ignored = StdioStream()
-    sys.stdout = ignored
-    sys.stderr = ignored
+def _replace_fd(file, target):
+    """
+    Temporarily replace the file descriptor for `file`,
+    for which sys.stdout or sys.stderr is passed.
+    """
+    try:
+        fd = file.fileno()
+    except AttributeError:
+        # `file` does not have fileno() so it's been replaced from the
+        # default sys.stdout, etc. Return with noop.
+        return
+    target_fd = target.fileno()
+
+    # `os.dup2()` closes the original FD, so we make copies.
+    dup_fd = os.dup(fd)
     try:
-        yield ignored
+        # Point the FD at the target.
+        os.dup2(target_fd, fd)
+        try:
+            yield
+        finally:
+            # Point the FD back at the original.
+            os.dup2(dup_fd, fd)
     finally:
-        sys.stdout = sys.__stdout__
-        sys.stderr = sys.__stderr__
+        os.close(dup_fd)
 
 
-if sys.version_info < (3,):
+@contextlib.contextmanager
+def _replace_stdx(attr, target):
+    assert attr in ("stdout", "stderr")
+    orig = getattr(sys, attr)
+    setattr(sys, attr, target)
+    try:
+        yield orig
+    finally:
+        setattr(sys, attr, orig)
 
-    class StdioStream(StringIO):
-        def write(self, msg):
-            StringIO.write(self, msg.decode())
 
+class _StringIO:
+    def setvalue(self, value):
+        self._value = value
 
-else:
-    StdioStream = StringIO
+    def getvalue(self):
+        return self._value
+
+
+@contextlib.contextmanager
+def _temp_io():
+    td = tempfile.mkdtemp()
+    path = os.path.join(td, "temp")
+    sio = _StringIO()
+    try:
+        with open(path, "w") as f:
+            yield sio, f
+    finally:
+        with open(path, "r") as f:
+            sio.setvalue(f.read())
+        try:
+            os.remove(path)
+        except OSError:
+            pass
+        os.rmdir(td)
```

Comment:
You could also use `tempfile.TemporaryFile`:

```suggestion
@contextlib.contextmanager
def _temp_io():
    sio = _StringIO()
    with tempfile.TemporaryFile() as tmp:
        try:
            yield sio, tmp
        finally:
            tmp.seek(0)
            sio.setvalue(tmp.read())
```

Reply:
Changed to "r+" mode.